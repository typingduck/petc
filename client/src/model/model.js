import axios from 'axios'
import io from 'socket.io-client'
import uuid from 'uuid/v4'

import config from '../conf/config'

export function initialDoc () {
  return {
    nodes: {},  // map of id -> {id, x, y}
    edges: {}   // map of id -> {id, source, target}
  }
}

export function createNode (x, y) {
  const id = uuid().substring(0, 8)
  return {id, x, y}
}

export function createEdge (source, target) {
  const id = source + '-' + target
  return {id, source, target}
}

export async function createNewDoc () {
  const {data} = await db().post('docs', initialDoc())
  return data
}

export async function loadDoc (docId) {
  const {data: doc} = await db().get(`docs/${docId}`)
  return removeMetaKeys(doc)
}

export async function saveDoc (doc) {
  const {data} = await db().put(`docs/${doc._id}`, doc)
  return data
}

export async function connectSocket (docId, cb) {
  const socket = await io(config.API_URL)
  socket.emit('doc-select', docId)
  socket.on('doc-patch', cb)

  return {
    addNode: node =>
      socket.emit('doc-patch', { 'op': 'add', 'path': `/nodes/${node.id}`, 'value': node }),
    updateNode: node =>
      socket.emit('doc-patch', { 'op': 'replace', 'path': `/nodes/${node.id}`, 'value': node }),
    removeNode: node =>
      socket.emit('doc-patch', { 'op': 'remove', 'path': `/nodes/${node.id}` }),
    addEdge: edge =>
      socket.emit('doc-patch', { 'op': 'add', 'path': `/edges/${edge.id}`, 'value': edge })
  }
}

function db () {
  return axios.create({
    baseURL: config.API_URL
  })
}

function removeMetaKeys (doc) {
  delete doc['_id']
  delete doc['_rev']
  return doc
}
