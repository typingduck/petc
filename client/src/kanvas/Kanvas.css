#petc-kanvas {
  bottom: 0;
  color: grey;
  font-family: "Comic Sans MS", "Comic Sans", cursive;
  left: 0;
  overflow: hidden;
  padding: 0;
  position: absolute;
  right: 0;
  top: 0;
  z-index: var(--z-index-kanvas);
}

.petc-edge-label {
  border-radius: 10px;
  background-color: rgba(255, 255, 255, 0.9);
}

.petc-node {
  border-radius: 20px;
  border: 10px solid grey;
  display: block;
  margin: auto;
  height: 16px;
  width: 16px;
}

.petc-node-selected {
  border-color: red;
}

.petc-node-label {
  text-align: center;
}
