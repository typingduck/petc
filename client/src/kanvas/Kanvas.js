import React from 'react'
import {jsPlumb} from 'jsplumb'

import './Kanvas.css'
import {createNode, createEdge} from '../model/model'
import {isInTrashCan} from '../controls/Trashcan'

const JSPLUMP_DEFAULTS = {
  Container: 'petc-kanvas',
  PaintStyle: {
    strokeWidth: 2,
    stroke: 'rgba(100, 100, 100, 0.5)'
  },
  Connector: [ 'Straight' ],
  Endpoint: [ 'Dot', { radius: 1 } ],
  EndpointStyle: { fill: 'grey' },
  Anchor: [ 'Continuous', { shape: 'Circle' } ]
}

/**
 * Show nodes as circles on the page.
 */
class Kanvas extends React.Component {
  constructor (props) {
    super(props)
    this.state = {}

    this.handleClick = this.handleClick.bind(this)
    this.selectNode = this.selectNode.bind(this)
  }

  componentDidMount () {
    window.j = jsPlumb.getInstance(JSPLUMP_DEFAULTS)
    this.setState({jsPlmb: window.j})
  }

  handleClick (ev) {
    if (this.props.controls.isNodeMode) {
      this.props.addNode(createNode(ev.pageX - 25, ev.pageY - 25,
        this.props.controls.selectedNodeClass))
    }
    this.setState({selectedNode: null})
  }

  selectNode (nodeId) {
    if (this.props.controls.isEdgeMode) {
      // If existing selected node then connect, otherwise select
      if (this.state.selectedNode && nodeId !== this.state.selectedNode) {
        this.props.addEdge(
          createEdge(this.state.selectedNode, nodeId, this.props.controls.selectedEdgeClass))
        this.setState({selectedNode: null})
      } else {
        this.setState({selectedNode: nodeId})
      }
    }
  }

  render () {
    const selectedNode = this.props.controls.isEdgeMode && this.state.selectedNode
    if (!this.props.doc.style) {
      this.props.doc.style = { nodes: {}, edges: {} }
    }
    const nodes = !this.state.jsPlmb ? null : Object.values(this.props.doc.nodes).map(node =>
      <Node
        key={node.id}
        node={node}
        {...this.props}
        jsPlmb={this.state.jsPlmb}
        selectNode={this.selectNode}
        selected={node.id === selectedNode}
        updateNode={this.props.updateNode}
      />
    )
    const edges = !this.state.jsPlmb ? null : Object.values(this.props.doc.edges).map(edge =>
      this.props.doc.nodes[edge.source] && this.props.doc.nodes[edge.target] &&
        <Edge
          key={edge.id + edge.className}
          edge={edge}
          {...this.props}
          jsPlmb={this.state.jsPlmb}
        />
    )

    return (
      <div id='petc-kanvas' onClick={this.handleClick}>
        {nodes}
        {edges}
      </div>
    )
  }
}

class Node extends React.Component {
  constructor (props) {
    super(props)
    this._nodeId = this.props.node.id
    this._isMounted = false
    this._textRotation = Math.round(10 * (Math.random() - 0.5))

    this.handleClick = this.handleClick.bind(this)
    this.onDragStart = this.onDragStart.bind(this)
    this.onDrag = this.onDrag.bind(this)
    this.onDragStop = this.onDragStop.bind(this)
  }

  componentDidMount () {
    const dragOptions = {
      'start': this.onDragStart,
      'drag': this.onDrag,
      'stop': this.onDragStop,
      containment: true
      // grid:[25,25]
    }
    this.props.jsPlmb.draggable(nw(this._nodeId), dragOptions)
    this.props.jsPlmb.setDraggable(nw(this._nodeId), this.props.controls.isNodeMode)
    this._isMounted = true
  }

  componentWillUnmount () {
    this.props.jsPlmb.removeAllEndpoints(this._nodeId)
  }

  onDragStart () { }

  onDrag (ev) {
    this.props.draggingNode({
      id: this._nodeId,
      x: ev.e.pageX,
      y: ev.e.pageY
    })
  }

  onDragStop (ev) {
    this.props.draggingNode(null)
    const [x, y] = ev.finalPos
    const newNode = Object.assign({}, this.props.node, {x, y})
    if (isInTrashCan(newNode)) {
      this.props.removeNode(newNode)
    } else {
      this.props.updateNode(newNode)
    }
  }

  handleClick (ev) {
    ev.stopPropagation()
    this.props.selectNode(this._nodeId)
  }

  render () {
    const x = this.props.node.x
    const y = this.props.node.y
    if (this._isMounted &&  // need to render the element once before jsplumb can see it
        !this.props.controls.dragNode) {
      // This only exists to force redraw when another client moves a node. Can
      // possibly be optimized to prevent so many calls.
      this.props.jsPlmb.repaint(nw(this._nodeId), {left: x, top: y})
      this.props.jsPlmb.setDraggable(nw(this._nodeId), this.props.controls.isNodeMode)
    }
    let className = 'petc-node'
    if (this.props.selected) {
      className += ' petc-node-selected'
    }
    if (this.props.node.className) {
      className += ' ' + this.props.node.className
    }
    const nodeStyle = this.props.doc.style.nodes[this.props.node.className]
    const textStyle = { WebkitTransform: `rotate(${this._textRotation}deg)` }
    return (
      <div
        id={nw(this._nodeId)}
        style={{ position: 'absolute', left: x, top: y }}
      >
        <div
          id={this._nodeId}
          onClick={this.handleClick}
          className={className}
          style={nodeStyle}
        />
        <div className='petc-node-label' style={textStyle}>{this.props.node.label}</div>
      </div>
    )
  }
}

class Edge extends React.Component {
  componentDidMount () {
    const edge = this.props.edge
    const edgeInfo = {
      source: edge.source,
      target: edge.target,
      overlays: []
    }
    const edgeClassName = edge.className || 'default'
    const optStyle = this.props.doc.style.edges[edgeClassName]
    if (optStyle) {
      Object.assign(edgeInfo, optStyle)
    }
    if (edge.label) {
      edgeInfo.overlays.push(
        [ 'Label', { label: edge.label, cssClass: 'petc-edge-label' } ]
      )
    }
    this._connection = this.props.jsPlmb.connect(edgeInfo)
  }

  componentWillUnmount () {
    try {
      this.props.jsPlmb.deleteConnection(this._connection)
    } catch (e) {
      // ignore if already removed
    }
  }

  render () {
    // could return null, but return something helps with testing
    return <div id={this.props.edge.id} />
  }
}

/**
 * Id of wrappper containing the node and its label.
 */
function nw (nodeId) {
  return nodeId + '-wrapper'
}

export default Kanvas
