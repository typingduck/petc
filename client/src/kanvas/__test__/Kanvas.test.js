/* global beforeEach expect it jest */
import React from 'react'
import Renderer from 'react-test-renderer'

import {initialDoc} from '../../model/model'
import Kanvas from '../Kanvas'

jest.useFakeTimers()

Math.random = () => 0.1

const mockJsPlumb = {
  connect: jest.fn(),
  draggable: jest.fn(),
  repaint: jest.fn(),
  setDraggable: jest.fn(),
  bind: jest.fn(),
  importDefaults: jest.fn()
}

jest.mock('jsplumb', () => (
  { jsPlumb: { getInstance: () => mockJsPlumb } }
))

let mockIsInTrashCan = false

jest.mock('../../controls/Trashcan', () => (
  { isInTrashCan: jest.fn(() => mockIsInTrashCan) }
))

jest.mock('uuid/v4', () => () => 'u-u-i-d')

beforeEach(() => {
  jest.clearAllMocks()
})

it('renders initial doc', () => {
  const props = {
    doc: initialDoc(),
    controls: {
      isNodeMode: false,
      isEdgeMode: false
    }
  }
  const component = Renderer.create(<Kanvas {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('render with nodes and edges', () => {
  const props = {
    doc: {
      nodes: {
        'a': {id: 'a', x: 1, y: 2},
        'b': {id: 'b', x: 2, y: 3}
      },
      edges: {
        'a-b': {id: 'a-b', source: 'a', target: 'b'}
      }
    },
    controls: {
      isNodeMode: false,
      isEdgeMode: false
    }
  }
  const component = Renderer.create(<Kanvas {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})

it('should create node on click', () => {
  const props = {
    doc: {
      nodes: {},
      edges: {}
    },
    controls: {
      isNodeMode: true,
      isEdgeMode: false
    },
    addNode: jest.fn()
  }
  const component = Renderer.create(<Kanvas {...props} />)
  const tree = component.toJSON()

  tree.props.onClick({ pageX: 125, pageY: 225 })

  expect(props.addNode).toHaveBeenCalledWith({ id: 'u-u-i-d', x: 100, y: 200 })
})

it('should select node in edge mode', () => {
  const props = {
    doc: {
      nodes: {
        'node': { id: 'node', x: 100, y: 100 }
      },
      edges: {}
    },
    controls: {
      isNodeMode: false,
      isEdgeMode: true
    },
    addNode: jest.fn()
  }
  // given kanvas with a node
  const component = Renderer.create(<Kanvas {...props} />)
  let tree = component.toJSON()
  let nodeDiv = tree.children[0].children[0]
  expect(nodeDiv.props.id).toEqual('node')
  expect(nodeDiv.props.className).toEqual('petc-node')

  // when clicking on node
  nodeDiv.props.onClick({ stopPropagation: () => null })

  // should mark node as selected
  tree = component.toJSON()
  nodeDiv = tree.children[0].children[0]
  expect(nodeDiv.props.className).toEqual('petc-node petc-node-selected')

  // when clicking on kanvas
  tree.props.onClick()
  tree = component.toJSON()
  nodeDiv = tree.children[0]

  // should deselect node
  tree = component.toJSON()
  nodeDiv = tree.children[0].children[0]
  expect(nodeDiv.props.className).toEqual('petc-node')
})

it('should disable dragging in edge view mode', () => {
  const props = {
    doc: {
      nodes: {
        'node': { id: 'node', x: 100, y: 100 }
      },
      edges: {}
    },
    controls: {
      isNodeMode: false,
      isEdgeMode: true
    },
    addNode: jest.fn()
  }
  // given kanvas with a node in edge mode
  const component = Renderer.create(<Kanvas {...props} />)
  let tree = component.toJSON()
  let nodeDiv = tree.children[0]
  expect(nodeDiv.props.id).toEqual('node-wrapper')

  // then dragging should be disabled
  const setDraggable = mockJsPlumb.setDraggable.mock.calls[0]
  expect(setDraggable).toEqual([ 'node-wrapper', false ])
})

it('should enable dragging in node view mode', () => {
  const props = {
    doc: {
      nodes: {
        'node': { id: 'node', x: 100, y: 100 }
      },
      edges: {}
    },
    controls: {
      isNodeMode: true,
      isEdgeMode: false
    },
    addNode: jest.fn()
  }
  // given kanvas with a node in edge mode
  const component = Renderer.create(<Kanvas {...props} />)
  let tree = component.toJSON()
  let nodeDiv = tree.children[0]
  expect(nodeDiv.props.id).toEqual('node-wrapper')

  // then dragging should be disabled
  const setDraggable = mockJsPlumb.setDraggable.mock.calls[0]
  expect(setDraggable).toEqual([ 'node-wrapper', true ])
})

it('should emit node position after dragging', () => {
  const props = {
    doc: {
      nodes: {
        'node': { id: 'node', x: 10, y: 20 }
      },
      edges: {}
    },
    controls: {
      isNodeMode: false,
      isEdgeMode: true
    },
    draggingNode: jest.fn(),
    updateNode: jest.fn()
  }
  // given kanvas with a node
  const component = Renderer.create(<Kanvas {...props} />)
  let tree = component.toJSON()
  let nodeDiv = tree.children[0]
  expect(nodeDiv.props.id).toEqual('node-wrapper')

  // given jsPlumb dragging setup
  const draggable = mockJsPlumb.draggable.mock.calls[0]
  expect(draggable[0]).toEqual('node-wrapper')
  const onDrag = draggable[1].drag
  const onDragStart = draggable[1].start
  const onDragStop = draggable[1].stop

  // when dragging node
  onDragStart()

  onDrag({ e: { pageX: 100, pageY: 200 } })
  expect(props.draggingNode).toHaveBeenCalledWith({ id: 'node', x: 100, y: 200 })

  onDragStop({ finalPos: [1000, 2000] })
  expect(props.draggingNode).toHaveBeenCalledWith(null)

  // should emit node position at end
  expect(props.updateNode).toHaveBeenCalledWith({ id: 'node', x: 1000, y: 2000 })
})

it('should delete node if in trashcan', () => {
  const props = {
    doc: {
      nodes: {
        'node-id': { id: 'node-id', x: 10, y: 20 }
      },
      edges: {}
    },
    controls: {
      isNodeMode: false,
      isEdgeMode: true
    },
    draggingNode: jest.fn(),
    updateNode: jest.fn(),
    removeNode: jest.fn()
  }
  // given kanvas with a node
  const component = Renderer.create(<Kanvas {...props} />)
  let tree = component.toJSON()
  let nodeDiv = tree.children[0]
  expect(nodeDiv.props.id).toEqual('node-id-wrapper')

  // given jsPlumb dragging setup
  const draggable = mockJsPlumb.draggable.mock.calls[0]
  expect(draggable[0]).toEqual('node-id-wrapper')
  const onDrag = draggable[1].drag
  const onDragStart = draggable[1].start
  const onDragStop = draggable[1].stop

  // when dragging node
  onDragStart()
  onDrag({ e: { pageX: 100, pageY: 200 } })

  mockIsInTrashCan = true
  onDragStop({ finalPos: [1000, 2000] })

  // should emit delete node
  expect(props.removeNode).toHaveBeenCalledWith('node-id')
})

it('should show node label', () => {
  const props = {
    doc: {
      nodes: {
        'node': { id: 'node', x: 10, y: 20, label: 'a label' }
      },
      edges: {}
    },
    controls: {
      isNodeMode: false,
      isEdgeMode: true
    }
  }
  // given kanvas with a node
  const component = Renderer.create(<Kanvas {...props} />)
  expect(component.toJSON()).toMatchSnapshot()
})
