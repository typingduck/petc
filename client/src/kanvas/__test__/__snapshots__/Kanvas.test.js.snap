// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render with nodes and edges 1`] = `
<div
  id="petc-kanvas"
  onClick={[Function]}
>
  <div
    id="a-wrapper"
    style={
      Object {
        "left": 1,
        "position": "absolute",
        "top": 2,
      }
    }
  >
    <div
      className="petc-node"
      id="a"
      onClick={[Function]}
      style={undefined}
    />
    <div
      className="petc-node-label"
      style={
        Object {
          "WebkitTransform": "rotate(-4deg)",
        }
      }
    />
  </div>
  <div
    id="b-wrapper"
    style={
      Object {
        "left": 2,
        "position": "absolute",
        "top": 3,
      }
    }
  >
    <div
      className="petc-node"
      id="b"
      onClick={[Function]}
      style={undefined}
    />
    <div
      className="petc-node-label"
      style={
        Object {
          "WebkitTransform": "rotate(-4deg)",
        }
      }
    />
  </div>
  <div
    id="a-b"
  />
</div>
`;

exports[`renders initial doc 1`] = `
<div
  id="petc-kanvas"
  onClick={[Function]}
/>
`;

exports[`should show node label 1`] = `
<div
  id="petc-kanvas"
  onClick={[Function]}
>
  <div
    id="node-wrapper"
    style={
      Object {
        "left": 10,
        "position": "absolute",
        "top": 20,
      }
    }
  >
    <div
      className="petc-node"
      id="node"
      onClick={[Function]}
      style={undefined}
    />
    <div
      className="petc-node-label"
      style={
        Object {
          "WebkitTransform": "rotate(-4deg)",
        }
      }
    >
      a label
    </div>
  </div>
</div>
`;
